File: ops/static/access.js
// access.js — who can do what.
//
// Roles are enumerated and NO ROLE IMPLIES ANOTHER: an admin who is not
// also a viewer cannot read the register. That has caught us twice, so the
// grid shows all four roles per entity as checkboxes rather than a single
// "level" that would imply a hierarchy the system does not have.
//
// Until this screen existed, granting a role meant running Python against
// the database. That is fine for one person and impossible for anyone else.

import { api, h, mount, stateMessage } from "./app.js";

let pending = null;

function when(ts) {
  if (!ts) return "\u2013";
  return new Date(ts * 1000).toISOString().slice(0, 10);
}

export async function render(root) {
  mount(root, stateMessage("Loading users", null, false));
  let data;
  try {
    data = await api("GET", "/api/users");
  } catch (err) {
    mount(root, err.status === 403
      ? stateMessage("Admin only",
          "Managing access needs the admin role on an entity.", false)
      : stateMessage("Could not load users", err.message, true));
    return;
  }

  const notice = h("div", { class: "notice", hidden: true });
  if (pending) {
    notice.textContent = pending;
    notice.hidden = false;
    pending = null;
  }
  const report = (message) => { pending = message; render(root); };

  async function toggle(user, entity, role, on) {
    try {
      await api(on ? "POST" : "DELETE", `/api/users/${user.id}/roles`,
                { entity_id: entity.id, role });
      report(`${user.display_name}: ${on ? "granted" : "revoked"} ${role}`);
    } catch (err) {
      report(err.message);
    }
  }

  async function setActive(user, active) {
    if (!active && !window.confirm(
        `Switch off ${user.display_name}? Every session they hold stops `
        + "working immediately.")) return;
    try {
      await api("PATCH", `/api/users/${user.id}`, { is_active: active });
      report(`${user.display_name}: ${active ? "reactivated" : "switched off"}`);
    } catch (err) {
      report(err.message);
    }
  }

  const held = (user, entityId, role) =>
    user.roles.some((r) => r.entity_id === entityId && r.role === role);

  function roleCell(user, entity, role) {
    const editable = data.administers.includes(entity.id) && user.is_active;
    const on = held(user, entity.id, role);
    if (!editable) {
      return h("td", { class: on ? "num" : "num zero" },
        on ? "\u25cf" : "\u2013");
    }
    const box = h("input", {
      type: "checkbox", checked: on,
      "aria-label": `${role} for ${user.display_name} on ${entity.name}`,
    });
    box.addEventListener("change", () => toggle(user, entity, role, box.checked));
    return h("td", { class: "num" }, box);
  }

  const rows = [];
  for (const user of data.users) {
    for (const entity of data.entities) {
      rows.push(h("tr", { class: user.is_active ? null : "muted" },
        h("td", { class: "text-wide", title: user.email },
          user.display_name,
          user.id === data.me ? h("span", { class: "tag" }, "you") : null,
          user.is_active ? null : h("span", { class: "tag is-off" }, "off")),
        h("td", { class: "muted text", title: user.email }, user.email),
        data.entities.length > 1
          ? h("td", { class: "muted" }, entity.name) : null,
        ...data.roles.map((role) => roleCell(user, entity, role)),
        h("td", { class: "mono muted" }, when(user.last_seen_ts)),
        h("td", null,
          data.administers.length && user.id !== data.me
            ? h("button", { type: "button",
                            class: user.is_active ? "danger" : null,
                            onclick: () => setActive(user, !user.is_active) },
                user.is_active ? "Switch off" : "Reactivate")
            : null)));
    }
  }

  mount(root, h("div", { class: "content" },
    h("div", { class: "page-head" },
      h("h1", null, "Access"),
      h("span", { class: "eyebrow" }, "roles by entity")),
    notice,
    h("div", { class: "table-wrap" },
      h("table", null,
        h("thead", null, h("tr", null,
          h("th", null, "User"),
          h("th", null, "Email"),
          data.entities.length > 1 ? h("th", null, "Entity") : null,
          ...data.roles.map((r) => h("th", { class: "num" }, r)),
          h("th", null, "Last seen"),
          h("th", null, ""))),
        h("tbody", null, rows))),
    // Said once, here, because the alternative is someone concluding the
    // system is broken when an admin cannot see the register.
    h("p", { class: "muted note" },
      "No role implies another. An admin who is not also a viewer cannot "
      + "read the register \u2014 grant every role the person needs.")));
}

File: ops/modules/access.py
"""Access module — who can do what (STP-1, §9).

Roles are enumerated and NO ROLE IMPLIES ANOTHER. An admin who is not also
a viewer cannot read the register, which has caught us twice and is
deliberate: the alternative is a hierarchy where granting one thing quietly
grants three.

Everything here is admin-only, and an admin cannot remove the last admin on
an entity -- including themselves. A system nobody can administer is one
that needs a database client to recover, which is exactly the situation
this screen exists to end.
"""

from typing import Any

from ops.db import Db
from ops.http_util import HttpError, Router

ROLE_ADMIN = "admin"
#: `finance` opens the office-expense screen -- the rent, the
#: subscriptions, the total cost of running the business. `payroll` is what
#: shows individual salaries, and it is a SEPARATE grant: somebody can see
#: that wages cost $96,250.01 in July without seeing what Justin earns.
#: Neither implies the other, and admin implies neither.
ROLES = ("viewer", "operations", "approver", "admin", "finance",
         "payroll")


def entity_ids(user: dict[str, Any]) -> list[int]:
    return sorted({r["entity_id"] for r in user["roles"]})


def admin_on(user: dict[str, Any], entity_id: int) -> bool:
    return any(r["role"] == ROLE_ADMIN and r["entity_id"] == entity_id
               for r in user["roles"])


def register(router: Router, db: Db) -> None:

    def require_admin(user, entity_id):
        if not admin_on(user, entity_id):
            raise HttpError(403, f"admin on entity {entity_id} is required")

    @router.route("/api/users", role=ROLE_ADMIN)
    def list_users(handler, user):
        ids = entity_ids(user)
        admin_ids = [e for e in ids if admin_on(user, e)]
        if not admin_ids:
            raise HttpError(403, "admin on an entity is required")
        users = db.users_with_roles()
        # Grants on entities this admin does not administer are shown but
        # not editable: hiding them would make a user look less privileged
        # than they are, which is the wrong way for that to be wrong.
        for row in users:
            for grant in row["roles"]:
                grant["editable"] = grant["entity_id"] in admin_ids
        # Only entities that are actually in use. The schema is
        # multi-entity from migration 001, but the interface stays
        # single-entity until a second one has something in it -- otherwise
        # this screen shows three rows per person for one real decision.
        # An entity appears the moment it gains a project or a grant.
        entities = db.query(
            """SELECT e.id, e.name FROM entity e
               WHERE EXISTS (SELECT 1 FROM project p WHERE p.entity_id = e.id)
                  OR EXISTS (SELECT 1 FROM user_entity_role r
                             WHERE r.entity_id = e.id)
               ORDER BY e.id""")
        if not entities:
            entities = db.query("SELECT id, name FROM entity ORDER BY id LIMIT 1")
        return 200, {
            "users": users,
            "roles": list(ROLES),
            "entities": entities,
            "administers": admin_ids,
            "me": user["id"],
        }

    @router.route("/api/users/{user_id}/roles", role=ROLE_ADMIN, method="POST")
    def grant(handler, user, user_id):
        payload = handler.read_json()
        errors = {}
        target = db.query_one("SELECT id, display_name FROM users WHERE id = ?",
                              (user_id,))
        if target is None:
            raise HttpError(404, "not found")
        role = (payload.get("role") or "").strip()
        if role not in ROLES:
            errors["role"] = f"must be one of {', '.join(ROLES)}"
        entity_id = db.scalar("SELECT id FROM entity WHERE id = ?",
                              (payload.get("entity_id"),))
        if entity_id is None:
            errors["entity_id"] = "required"
        if errors:
            raise HttpError(400, "validation failed", errors)
        require_admin(user, entity_id)
        db.grant_role(target["id"], entity_id, role, user["id"])
        return 200, {"granted": role, "entity_id": entity_id,
                     "user": target["display_name"]}

    @router.route("/api/users/{user_id}/roles", role=ROLE_ADMIN, method="DELETE")
    def revoke(handler, user, user_id):
        payload = handler.read_json()
        target = db.query_one("SELECT id, display_name FROM users WHERE id = ?",
                              (user_id,))
        if target is None:
            raise HttpError(404, "not found")
        role = (payload.get("role") or "").strip()
        entity_id = db.scalar("SELECT id FROM entity WHERE id = ?",
                              (payload.get("entity_id"),))
        if role not in ROLES or entity_id is None:
            raise HttpError(400, "validation failed",
                            {"role": "unknown", "entity_id": "unknown"})
        require_admin(user, entity_id)
        if role == ROLE_ADMIN:
            remaining = [uid for uid in db.admins_on(entity_id)
                         if uid != target["id"]]
            if not remaining:
                # Including yourself. A system nobody can administer needs a
                # database client to recover.
                raise HttpError(
                    409, "that is the last admin on this entity; grant "
                         "another before removing this one")
        db.revoke_role(target["id"], entity_id, role, user["id"])
        return 200, {"revoked": role, "entity_id": entity_id,
                     "user": target["display_name"]}

    @router.route("/api/users/{user_id}", role=ROLE_ADMIN, method="PATCH")
    def set_active(handler, user, user_id):
        payload = handler.read_json()
        target = db.query_one(
            "SELECT id, display_name, is_active FROM users WHERE id = ?",
            (user_id,))
        if target is None:
            raise HttpError(404, "not found")
        if "is_active" not in payload:
            raise HttpError(400, "validation failed",
                            {"is_active": "required"})
        active = bool(payload["is_active"])
        # Only an admin of an entity the user has a role on, so an admin of
        # one company cannot switch off someone who works for another.
        theirs = {g["entity_id"] for g in db.query(
            "SELECT entity_id FROM user_entity_role WHERE user_id = ?",
            (target["id"],))}
        mine = {e for e in entity_ids(user) if admin_on(user, e)}
        if theirs and not (theirs & mine):
            raise HttpError(403, "this user has no role on an entity you "
                                 "administer")
        if not active:
            if target["id"] == user["id"]:
                raise HttpError(409, "you cannot switch off your own account")
            for entity_id in theirs & mine:
                remaining = [uid for uid in db.admins_on(entity_id)
                             if uid != target["id"]]
                if not remaining:
                    raise HttpError(
                        409, f"that is the last admin on entity {entity_id}")
        changed = db.set_user_active(target["id"], active, user["id"])
        return 200, {"changed": changed, "is_active": active,
                     "user": target["display_name"]}
